### Business Understanding

##### GOAL: To detect the Fake reviews

#### 0.Data Loading <a id=dc>
Back to Top 

In [6]:
#  Data manipulation Libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

ModuleNotFoundError: No module named 'pandas'

In [ ]:
# loading the data
data=pd.read_csv(r"C:\Users\91905\Downloads\fake reviews dataset.csv")

In [ ]:
data

category  rating label  \
0                Home_and_Kitchen_5     5.0    CG   
1                Home_and_Kitchen_5     5.0    CG   
2                Home_and_Kitchen_5     5.0    CG   
3                Home_and_Kitchen_5     1.0    CG   
4                Home_and_Kitchen_5     5.0    CG   
...                             ...     ...   ...   
40427  Clothing_Shoes_and_Jewelry_5     4.0    OR   
40428  Clothing_Shoes_and_Jewelry_5     5.0    CG   
40429  Clothing_Shoes_and_Jewelry_5     2.0    OR   
40430  Clothing_Shoes_and_Jewelry_5     1.0    CG   
40431  Clothing_Shoes_and_Jewelry_5     5.0    OR   

                                                   text_  
0      Love this!  Well made, sturdy, and very comfor...  
1      love it, a great upgrade from the original.  I...  
2      This pillow saved my back. I love the look and...  
3      Missing information on how to use it, but it i...  
4      Very nice set. Good quality. We have had the s...  
...                                                  ...  
40427  I had read some reviews saying that this bra r...  
40428  I wasn't sure exactly what it would be. It is ...  
40429  You can wear the hood by itself, wear it with ...  
40430  I liked nothing about this dress. The only rea...  
40431  I work in the wedding industry and have to wor...  

[40432 rows x 4 columns]

This data set is containing 40432 rows and 4 columns

**Basic cheks of the data**

In [ ]:
display("First five rows of Data:",data.head())
print()
display("Last five rows of Data:",data.tail())

'First five rows of Data:'

category  rating label  \
0  Home_and_Kitchen_5     5.0    CG   
1  Home_and_Kitchen_5     5.0    CG   
2  Home_and_Kitchen_5     5.0    CG   
3  Home_and_Kitchen_5     1.0    CG   
4  Home_and_Kitchen_5     5.0    CG   

                                               text_  
0  Love this!  Well made, sturdy, and very comfor...  
1  love it, a great upgrade from the original.  I...  
2  This pillow saved my back. I love the look and...  
3  Missing information on how to use it, but it i...  
4  Very nice set. Good quality. We have had the s...

'Last five rows of Data:'

category  rating label  \
40427  Clothing_Shoes_and_Jewelry_5     4.0    OR   
40428  Clothing_Shoes_and_Jewelry_5     5.0    CG   
40429  Clothing_Shoes_and_Jewelry_5     2.0    OR   
40430  Clothing_Shoes_and_Jewelry_5     1.0    CG   
40431  Clothing_Shoes_and_Jewelry_5     5.0    OR   

                                                   text_  
40427  I had read some reviews saying that this bra r...  
40428  I wasn't sure exactly what it would be. It is ...  
40429  You can wear the hood by itself, wear it with ...  
40430  I liked nothing about this dress. The only rea...  
40431  I work in the wedding industry and have to wor...

In [ ]:
data.shape

(40432, 4)

In [ ]:
display("Missing value and data types of each column:",data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40432 entries, 0 to 40431
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   category  40432 non-null  object 
 1   rating    40432 non-null  float64
 2   label     40432 non-null  object 
 3   text_     40432 non-null  object 
dtypes: float64(1), object(3)
memory usage: 1.2+ MB


'Missing value and data types of each column:'

None

In [ ]:
display("Missing value count:",data.isnull().sum())

'Missing value count:'

category    0
rating      0
label       0
text_       0
dtype: int64

In [ ]:
# converting object column data into lower case
for i in data.select_dtypes("O"):
    data[i]=data[i].str.lower()

In [ ]:
# duplicate rows checking
data[data.duplicated()]

category  rating label  \
6025          sports_and_outdoors_5     5.0    cg   
6708          sports_and_outdoors_5     5.0    cg   
12548               movies_and_tv_5     5.0    cg   
19802                pet_supplies_5     5.0    cg   
20242                pet_supplies_5     5.0    cg   
22305                pet_supplies_5     5.0    or   
26444                kindle_store_5     5.0    cg   
27233                kindle_store_5     5.0    cg   
29203                       books_5     5.0    cg   
33602              toys_and_games_5     5.0    cg   
34795              toys_and_games_5     5.0    cg   
37543  clothing_shoes_and_jewelry_5     5.0    cg   

                                                   text_  
6025   this is a really good starter kit, with lots o...  
6708   great, no complaints. comfortable, phone fits ...  
12548  one of the best movies of the year.  not for e...  
19802  my dog loves these and it has kept her occupie...  
20242  my dog loves it and it has kept her occupied f...  
22305  got these to give to my 8 mth old chihuahua wh...  
26444  i received this story as an arc in exchange fo...  
27233  this is the first book in a series by the auth...  
29203  i really enjoyed this book. the characters wer...  
33602  i got this for my son for christmas.  he loved...  
34795  i bought this for my son for christmas.  he lo...  
37543  i bought this for a halloween costume and it w...

In [ ]:
# dropping the duplicated rows
data=data.drop_duplicates().reset_index(drop=True)

In [ ]:
data[data.duplicated()]

Empty DataFrame
Columns: [category, rating, label, text_]
Index: []

#### 1. Text cleaning

Removing punctuations and other special characters

In [ ]:
data.columns

Index(['category', 'rating', 'label', 'text_'], dtype='object')

In [ ]:
# Renaming the column name
data.rename({"text_":'text'},axis=1,inplace=True)

In [ ]:
data.columns

Index(['category', 'rating', 'label', 'text'], dtype='object')

In [ ]:
data.category.unique()

array(['home_and_kitchen_5', 'sports_and_outdoors_5', 'electronics_5',
       'movies_and_tv_5', 'tools_and_home_improvement_5',
       'pet_supplies_5', 'kindle_store_5', 'books_5', 'toys_and_games_5',
       'clothing_shoes_and_jewelry_5'], dtype=object)

In [ ]:
data.category=data.category.str.replace('_'," ").str.strip('5')

In [ ]:
data.category.unique()

array(['home and kitchen ', 'sports and outdoors ', 'electronics ',
       'movies and tv ', 'tools and home improvement ', 'pet supplies ',
       'kindle store ', 'books ', 'toys and games ',
       'clothing shoes and jewelry '], dtype=object)

In [ ]:
data.rating.unique()

array([5., 1., 3., 2., 4.])

In [ ]:
data.label.unique()

array(['cg', 'or'], dtype=object)

In [ ]:
data.text.unique()

array(['love this!  well made, sturdy, and very comfortable.  i love it!very pretty',
       "love it, a great upgrade from the original.  i've had mine for a couple of years",
       'this pillow saved my back. i love the look and feel of this pillow.',
       ...,
       'you can wear the hood by itself, wear it with the hood, or wear just the jacket without the hood.  how is that for 3-in-1 systems?\n\nreminiscent of what lands end sells as a "squall jacket", or what uniform and corporate wear sellers call a three season jacket.\n\nwater and wind resistant shell.  microfleece lined body.  smooth polyester lining in sleeves.  left hand side, napoleon style, inside pocket with a hook and loop closure.  adjustable sleeve cuffs.  hemline cinch cord with toggles.  molded plastic chain style, off brand, slide fastener.\n\npolyester insulation is spilling out due to poor quality stitching.  item was delivered with a broken main zipper.\n\ni can wear this jacket over just a t-shirt, and exp

In [ ]:
import re
import nltk

In [ ]:
text_cleaning_pattern= "@\S+|https?:\S+|http?:\S|[^A-Za-z0-9]+"

In [ ]:
data['text']=data['text'].apply((lambda x:re.sub(text_cleaning_pattern,' ',x)))

In [ ]:
data.text[6]

'they are the perfect touch for me and the only thing i wish they had a little more space '

#### 2.Text Preprocessing for Modelling

* **Tokenization**

In [ ]:
from keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [ ]:
tokenizer=Tokenizer()
tokenizer.fit_on_texts(data['text'])
X=tokenizer.texts_to_sequences(data['text'])

In [ ]:
tokenizer.word_index

{'the': 1,
 'i': 2,
 'a': 3,
 'and': 4,
 'it': 5,
 'to': 6,
 'is': 7,
 'this': 8,
 'of': 9,
 'for': 10,
 'in': 11,
 'that': 12,
 'was': 13,
 'my': 14,
 'with': 15,
 'have': 16,
 'very': 17,
 'but': 18,
 's': 19,
 'not': 20,
 'are': 21,
 'great': 22,
 'on': 23,
 'you': 24,
 'book': 25,
 't': 26,
 'good': 27,
 'one': 28,
 'they': 29,
 'as': 30,
 'so': 31,
 'be': 32,
 'love': 33,
 'well': 34,
 'had': 35,
 'will': 36,
 'would': 37,
 'story': 38,
 'like': 39,
 'read': 40,
 'just': 41,
 'has': 42,
 'these': 43,
 'he': 44,
 'she': 45,
 'them': 46,
 'can': 47,
 'if': 48,
 'little': 49,
 'an': 50,
 'all': 51,
 'more': 52,
 'about': 53,
 'only': 54,
 'her': 55,
 'get': 56,
 'from': 57,
 'recommend': 58,
 'up': 59,
 'characters': 60,
 'really': 61,
 'at': 62,
 'am': 63,
 'who': 64,
 'use': 65,
 'out': 66,
 'or': 67,
 'bought': 68,
 'also': 69,
 'movie': 70,
 'too': 71,
 'we': 72,
 'm': 73,
 'nice': 74,
 'me': 75,
 'time': 76,
 'small': 77,
 'when': 78,
 'easy': 79,
 'what': 80,
 'there': 81,
 'ot

In [ ]:
%time
vocab_size=len(tokenizer.word_index)+1
print("Total words:" , vocab_size)

CPU times: total: 0 ns
Wall time: 0 ns
Total words: 40980


* **Padding**

In [ ]:
for i in data.text:
    textlengths=len(i.split())

In [ ]:
textlengths

355

In [ ]:
np.mean(textlengths)

355.0

In [ ]:
# Length of text to consider
maxlengthtext=300
X=pad_sequences(X,maxlengthtext)
print(X.shape)

(40420, 300)


In [ ]:
X[0]

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   

#### 3.Word embeddings

In [ ]:
import gensim

* **Defining Model**

In [ ]:
# word2vec Parameters
W2V_SIZE=300
W2V_WINDOW=5
W2V_EPOCH=32
W2V_MIN_COUNT=1

In [ ]:
w2v_model = gensim.models.word2vec.Word2Vec(vector_size=W2V_SIZE, window=W2V_WINDOW, min_count=W2V_MIN_COUNT, workers=8)

In [ ]:
%%time
document_with_words=[_text.split() for _text in data.text]

CPU times: total: 547 ms
Wall time: 701 ms


In [ ]:
len(document_with_words)

40420

* **Vocabulary Building**

In [ ]:
w2v_model.build_vocab(document_with_words)

In [ ]:
wv_words=list(w2v_model.wv.index_to_key)
wv_vocab_size=len(wv_words)
print("W2Vec Vocab size:",wv_vocab_size)

W2Vec Vocab size: 40979


* **Training Word2Vec for Word Embeddings**

In [ ]:
%%time
w2v_model.train(document_with_words,total_examples=len(document_with_words),epochs=W2V_EPOCH)

CPU times: total: 6min 11s
Wall time: 1min 31s


(62478270, 89727968)

In [ ]:
w2v_model.wv.get_vector('online')

array([-0.9343838 ,  1.5417063 , -0.3092251 , -1.5874343 ,  0.5736539 ,
        0.34915873,  1.4758911 ,  0.19531612,  0.71648306,  0.28973463,
        0.46682614,  0.71165735,  0.3047982 , -0.2110059 , -0.07866424,
       -0.93094593,  0.97046053, -0.90169924, -1.4323399 , -0.24344048,
       -0.6603118 ,  0.28562415, -0.22421376,  0.97499347, -0.66931117,
        1.5591447 ,  0.921768  , -1.1698439 ,  1.2183344 , -0.09755536,
       -2.0823445 , -0.19440958, -1.0400598 , -0.81461364, -0.09823321,
       -0.28345183,  0.92832416,  0.09325775, -0.6817446 ,  1.6165243 ,
        0.28001896,  0.23936729,  1.2605183 , -1.3552829 , -1.3358922 ,
        1.0157673 , -0.7905675 ,  0.58255285, -1.3915012 ,  0.11025988,
       -0.7920746 ,  0.9656289 , -0.8757637 ,  0.77800876, -3.0521824 ,
        0.74123454,  0.66779745,  0.7231871 ,  0.32113972,  0.1382144 ,
       -0.39978015, -0.6001611 ,  0.09575054, -1.3618351 ,  1.1823379 ,
       -0.26061887,  0.49568132,  0.2861932 ,  0.6140619 ,  1.75

#### 4.LSTM Model Building

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from keras.utils import to_categorical
from keras.models import Sequential
from keras.layers import Activation, Dense, Dropout, Embedding,LSTM
from keras.callbacks import ReduceLROnPlateau , EarlyStopping
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [ ]:
y=data['label']

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=123)

In [ ]:
x_train

array([[   0,    0,    0, ...,   96, 1863,   18],
       [   0,    0,    0, ...,    2,   37,   58],
       [   0,    0,    0, ...,  205,   10,   46],
       ...,
       [   0,    0,    0, ...,    3,  127,  111],
       [   0,    0,    0, ...,   50,  267,  160],
       [   0,    0,    0, ...,  109,   30,  798]])

In [ ]:
le=LabelEncoder()
y_train=le.fit_transform(y_train)

In [ ]:
y_train=to_categorical(y_train,num_classes=2,dtype='float32')

* **Embedding Layer Design**

In [ ]:
embedding_matrix=np.zeros((vocab_size,W2V_SIZE))
for word , i in tokenizer.word_index.items():
    if word in w2v_model.wv:
        embedding_matrix[i]=w2v_model.wv[word]
print(embedding_matrix.shape)

(40980, 300)


In [ ]:
embedding_layer=Embedding(vocab_size,W2V_SIZE,weights=[embedding_matrix],input_length=maxlengthtext,trainable=False)

* **LSTM Model**

In [ ]:
lstm=Sequential()
lstm.add(embedding_layer)
lstm.add(Dropout(0.5))
lstm.add(LSTM(100,dropout=0.2,recurrent_dropout=0.2))
lstm.add(Dense(2,activation='sigmoid'))
lstm.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 300, 300)          12294000  
                                                                 
 dropout (Dropout)           (None, 300, 300)          0         
                                                                 
 lstm (LSTM)                 (None, 100)               160400    
                                                                 
 dense (Dense)               (None, 2)                 202       
                                                                 
Total params: 12454602 (47.51 MB)
Trainable params: 160602 (627.35 KB)
Non-trainable params: 12294000 (46.90 MB)
_________________________________________________________________


In [8]:
lstm.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

NameError: name 'lstm' is not defined

In [ ]:
callbacks=[ReduceLROnPlateau(monitor='val_loss',patience=5,cooldown=0),EarlyStopping(monitor='val_accuracy',min_delta=0.001,patience=5,verbose=1)]

* **Training**

In [ ]:
%%time
history = lstm.fit(x_train, y_train,batch_size=10,epochs=10,validation_split=0.2,verbose=1,callbacks=callbacks,workers=2) 

NameError: name 'lstm' is not defined

* **Evaluation**

In [ ]:
%%time
y_act = le.transform(y_test)
y_act = to_categorical(y_act)
score = lstm.evaluate(x_test, y_act, batch_size=10)
print()
print("ACCURACY:",score[1])
print("LOSS:",score[0]) 

NameError: name 'le' is not defined

In [ ]:
acc = history.history['accuracy']
val_acc = history.history['val_accuracy']
loss = history.history['loss']
val_loss = history.history['val_loss']
epochs = range(len(acc))
plt.plot(epochs, acc, 'b', label='Training acc')
plt.plot(epochs, val_acc, 'r', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()
plt.figure()
plt.plot(epochs, loss, 'b', label='Training loss')
plt.plot(epochs, val_loss, 'r', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()
plt.show()

In [ ]:
%%time
y_pred = []
scores = lstm.predict(x_test, verbose=1, batch_size=5)
for i in range(len(scores)):
    y_pred.append(np.argmax(scores[i]))

#### Real Time Prediction

In [ ]:
labels={0:'Original_Review',1:'Computer_Generated_Review'}

In [ ]:
def predict(text):
    import time
    start_at = time.time()
    # Tokenize text
    x_test = pad_sequences(tokenizer.texts_to_sequences([text]), maxlen=300)
    # Predict
    score = lstm.predict([x_test])
    cl = np.argmax(score)
    # Decode Prediction
    label = labels[cl]
    return {"prediction": cl, "label": label,"elapsed_time": time.time()-start_at} 

In [ ]:
predict('love this well made sturdy and very comfortable i love it very pretty')